# 빅분기 실기 유형 연습

In [1]:
# 여러줄 출력 가능
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

# 이직여부 판단 데이터
  
데이터 설명 : 이직여부 판단 데이터 (target: 1: 이직 , 0 : 이직 x)  
x_train: https://raw.githubusercontent.com/Datamanim/datarepo/main/HRdata/X_train.csv  
y_train: https://raw.githubusercontent.com/Datamanim/datarepo/main/HRdata/y_train.csv  
x_test: https://raw.githubusercontent.com/Datamanim/datarepo/main/HRdata/X_test.csv  
x_label(평가용) : https://raw.githubusercontent.com/Datamanim/datarepo/main/HRdata/y_test.csv  
데이터 출처 :https://www.kaggle.com/datasets/arashnic/hr-analytics-job-change-of-data-scientists (참고, 데이터 수정)  

In [8]:
import pandas as pd
#데이터 로드
x_train = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/HRdata/X_train.csv")
y_train = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/HRdata/y_train.csv")
x_test= pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/HRdata/X_test.csv")

display(x_train.head())
display(y_train.head())

enrollee_id      city  city_development_index gender  \
0        25298  city_138                   0.836   Male   
1         4241  city_160                   0.920   Male   
2        24086   city_57                   0.866   Male   
3        26773   city_16                   0.910   Male   
4        32325  city_143                   0.740    NaN   

       relevent_experience enrolled_university education_level  \
0   No relevent experience    Full time course     High School   
1   No relevent experience    Full time course     High School   
2   No relevent experience       no_enrollment        Graduate   
3  Has relevent experience       no_enrollment        Graduate   
4   No relevent experience    Full time course        Graduate   

  major_discipline experience company_size company_type last_new_job  \
0              NaN          5      100-500      Pvt Ltd            1   
1              NaN          5          NaN          NaN            1   
2             STEM         10          NaN          NaN            1   
3             STEM        >20        50-99      Pvt Ltd           >4   
4             STEM          5          NaN          NaN        never   

   training_hours  
0              45  
1              17  
2              50  
3             135  
4              17

enrollee_id  target
0        25298     0.0
1         4241     1.0
2        24086     0.0
3        26773     0.0
4        32325     1.0

In [3]:
x_train.info()
x_test.info()
y_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12452 entries, 0 to 12451
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   enrollee_id             12452 non-null  int64  
 1   city                    12452 non-null  object 
 2   city_development_index  12452 non-null  float64
 3   gender                  9535 non-null   object 
 4   relevent_experience     12452 non-null  object 
 5   enrolled_university     12195 non-null  object 
 6   education_level         12137 non-null  object 
 7   major_discipline        10586 non-null  object 
 8   experience              12415 non-null  object 
 9   company_size            8600 non-null   object 
 10  company_type            8471 non-null   object 
 11  last_new_job            12179 non-null  object 
 12  training_hours          12452 non-null  int64  
dtypes: float64(1), int64(2), object(10)
memory usage: 1.2+ MB
<class 'pandas.core.frame.DataFra

In [6]:
x_train.isnull().sum()
x_test.isnull().sum()
y_train.isnull().sum()

enrollee_id                  0
city                         0
city_development_index       0
gender                    2917
relevent_experience          0
enrolled_university        257
education_level            315
major_discipline          1866
experience                  37
company_size              3852
company_type              3981
last_new_job               273
training_hours               0
dtype: int64

enrollee_id                  0
city                         0
city_development_index       0
gender                    1591
relevent_experience          0
enrolled_university        129
education_level            145
major_discipline           947
experience                  28
company_size              2086
company_type              2159
last_new_job               150
training_hours               0
dtype: int64

enrollee_id    0
target         0
dtype: int64

In [10]:
# print(x_train.info())
print(x_train.nunique())
# print(x_train.isnull().sum())
# 결측치가 있지만 따로 처리하지 않고 더미화


enrollee_id               12452
city                        122
city_development_index       92
gender                        3
relevent_experience           2
enrolled_university           3
education_level               5
major_discipline              6
experience                   22
company_size                  8
company_type                  6
last_new_job                  6
training_hours              241
dtype: int64


In [16]:
# 범주형 변수인데 적당히 많은 unique값을 가진 컬럼은 날린다.
drop_col = ['enrollee_id','city','company_type','experience']
x_train_drop = x_train.drop(columns = drop_col)
x_test_drop = x_test.drop(columns = drop_col)

In [17]:
# import sklearn
# print(sklearn.__all__)
# import sklearn.model_selection
# print(dir(sklearn.model_selection))

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
x_train_dummies = pd.get_dummies(x_train_drop)
y = y_train['target'].astype('int')

x_test_dummies = pd.get_dummies(x_test_drop)
# train과 컬럼 순서 동일하게 하기 (더미화 하면서 순서대로 정렬을 이미 하기 때문에 오류가 난다면 해당 컬럼이 누락된것)
x_test_dummies = x_test_dummies[x_train_dummies.columns]
# print(help(train_test_split))


In [18]:
X_train, X_validation, Y_train, Y_validation = train_test_split(x_train_dummies, y, test_size=0.33, random_state=42)
X_train.shape, X_validation.shape, Y_train.shape, Y_validation.shape

((8342, 35), (4110, 35), (8342,), (4110,))

In [19]:
rf = RandomForestClassifier(random_state =25)
rf.fit(X_train,Y_train)

RandomForestClassifier(random_state=25)

In [20]:
# import sklearn.metrics
# print(dir(sklearn.metrics))

from sklearn.metrics import accuracy_score , f1_score, recall_score, roc_auc_score ,precision_score

# model_score
predict_train_label = rf.predict(X_train)
predict_train_proba = rf.predict_proba(X_train)[:,1]

predict_validation_label = rf.predict(X_validation)
predict_validation_prob = rf.predict_proba(X_validation)[:,1]


In [21]:
# 문제에서 묻는 것에 따라 모델 성능 확인하기
# 정확도 (accuracy) , f1_score , recall , precision -> model.predict로 결과뽑기
# auc , 확률이라는 표현있으면 model.predict_proba로 결과뽑고 첫번째 행의 값을 가져오기 model.predict_proba()[:,1]
print('train accuracy :', accuracy_score(Y_train,predict_train_label))
print('validation accuracy :', accuracy_score(Y_validation,predict_validation_label))
print('\n')
print('train f1_score :', f1_score(Y_train,predict_train_label))
print('validation f1_score :', f1_score(Y_validation,predict_validation_label))
print('\n')
print('train recall_score :', recall_score(Y_train,predict_train_label))
print('validation recall_score :', recall_score(Y_validation,predict_validation_label))
print('\n')
print('train precision_score :', precision_score(Y_train,predict_train_label))
print('validation precision_score :', precision_score(Y_validation,predict_validation_label))
print('\n')
print('train auc :', roc_auc_score(Y_train,predict_train_proba))
print('validation auc :', roc_auc_score(Y_validation,predict_validation_prob))


train accuracy : 0.9965236154399425
validation accuracy : 0.7557177615571776


train f1_score : 0.992966286684453
validation f1_score : 0.4346846846846847


train recall_score : 0.9917635658914729
validation recall_score : 0.37079731027857826


train precision_score : 0.9941719281204469
validation precision_score : 0.5251700680272109


train auc : 0.9999187747418696
validation auc : 0.7437715758809


In [22]:
# test데이터 마찬가지 위와 같은 방식
predict_test_label = rf.predict(x_test_dummies)
predict_test_proba = rf.predict_proba(x_test_dummies)[:,1]

# accuracy, f1_score, recall, precision 
#pd.DataFrame({'enrollee_id': x_test.enrollee_id, 'target': predict_test_label}).to_csv('003000000.csv', index=False)

# auc, 확률
#pd.DataFrame({'enrollee_id': x_test.enrollee_id, 'target': predict_test_proba}).to_csv('003000000.csv', index=False)

# case 2

In [153]:
import pandas as pd
#데이터 로드
x_train = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/HRdata/X_train.csv")
y_train = pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/HRdata/y_train.csv")
x_test= pd.read_csv("https://raw.githubusercontent.com/Datamanim/datarepo/main/HRdata/X_test.csv")

display(x_train.head())
display(y_train.head())

enrollee_id      city  city_development_index gender  \
0        25298  city_138                   0.836   Male   
1         4241  city_160                   0.920   Male   
2        24086   city_57                   0.866   Male   
3        26773   city_16                   0.910   Male   
4        32325  city_143                   0.740    NaN   

       relevent_experience enrolled_university education_level  \
0   No relevent experience    Full time course     High School   
1   No relevent experience    Full time course     High School   
2   No relevent experience       no_enrollment        Graduate   
3  Has relevent experience       no_enrollment        Graduate   
4   No relevent experience    Full time course        Graduate   

  major_discipline experience company_size company_type last_new_job  \
0              NaN          5      100-500      Pvt Ltd            1   
1              NaN          5          NaN          NaN            1   
2             STEM         10          NaN          NaN            1   
3             STEM        >20        50-99      Pvt Ltd           >4   
4             STEM          5          NaN          NaN        never   

   training_hours  
0              45  
1              17  
2              50  
3             135  
4              17

enrollee_id  target
0        25298     0.0
1         4241     1.0
2        24086     0.0
3        26773     0.0
4        32325     1.0

In [154]:
# 공통 전처리 대상 데이터프레임 합치기 (동일한 처리를 위해)
train = x_train.copy()
test = x_test.copy()
all_data = pd.concat([train, test], axis=0, ignore_index=True)
all_data.head()
all_data.info()

enrollee_id      city  city_development_index gender  \
0        25298  city_138                   0.836   Male   
1         4241  city_160                   0.920   Male   
2        24086   city_57                   0.866   Male   
3        26773   city_16                   0.910   Male   
4        32325  city_143                   0.740    NaN   

       relevent_experience enrolled_university education_level  \
0   No relevent experience    Full time course     High School   
1   No relevent experience    Full time course     High School   
2   No relevent experience       no_enrollment        Graduate   
3  Has relevent experience       no_enrollment        Graduate   
4   No relevent experience    Full time course        Graduate   

  major_discipline experience company_size company_type last_new_job  \
0              NaN          5      100-500      Pvt Ltd            1   
1              NaN          5          NaN          NaN            1   
2             STEM         10          NaN          NaN            1   
3             STEM        >20        50-99      Pvt Ltd           >4   
4             STEM          5          NaN          NaN        never   

   training_hours  
0              45  
1              17  
2              50  
3             135  
4              17

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19158 entries, 0 to 19157
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   enrollee_id             19158 non-null  int64  
 1   city                    19158 non-null  object 
 2   city_development_index  19158 non-null  float64
 3   gender                  14650 non-null  object 
 4   relevent_experience     19158 non-null  object 
 5   enrolled_university     18772 non-null  object 
 6   education_level         18698 non-null  object 
 7   major_discipline        16345 non-null  object 
 8   experience              19093 non-null  object 
 9   company_size            13220 non-null  object 
 10  company_type            13018 non-null  object 
 11  last_new_job            18735 non-null  object 
 12  training_hours          19158 non-null  int64  
dtypes: float64(1), int64(2), object(10)
memory usage: 1.9+ MB


In [155]:
# 결측치 처리

# 1. gender
all_data['gender'] = all_data['gender'].fillna('Unknown')

# 2. enrolled_university
all_data['enrolled_university'] = all_data['enrolled_university'].fillna('no_enrollment')

# 3. education_level
all_data['education_level'] = all_data['education_level'].fillna('Unknown')

# 4. major_discipline
all_data['major_discipline'] = all_data['major_discipline'].fillna('No Major')


In [156]:
# 5. experience (문자 -> 숫자 변환 후 처리)
def clean_experience(val):
    if val == '>20':
        return 21
    elif val == '<1':
        return 0
    else:
        try:
            return int(val)
        except:
            return None

all_data['experience'] = all_data['experience'].apply(clean_experience)
all_data['experience'] = all_data['experience'].fillna(all_data['experience'].median())

all_data['experience']

0         5.0
1         5.0
2        10.0
3        21.0
4         5.0
         ... 
19153     8.0
19154     5.0
19155    19.0
19156     5.0
19157     5.0
Name: experience, Length: 19158, dtype: float64

In [157]:
# 6. company_size
all_data['company_size'] = all_data['company_size'].fillna('Unknown')

# 7. company_type
all_data['company_type'] = all_data['company_type'].fillna('Unknown')

all_data.head(3)

enrollee_id      city  city_development_index gender  \
0        25298  city_138                   0.836   Male   
1         4241  city_160                   0.920   Male   
2        24086   city_57                   0.866   Male   

      relevent_experience enrolled_university education_level  \
0  No relevent experience    Full time course     High School   
1  No relevent experience    Full time course     High School   
2  No relevent experience       no_enrollment        Graduate   

  major_discipline  experience company_size company_type last_new_job  \
0         No Major         5.0      100-500      Pvt Ltd            1   
1         No Major         5.0      Unknown      Unknown            1   
2             STEM        10.0      Unknown      Unknown            1   

   training_hours  
0              45  
1              17  
2              50

In [158]:
# 8. last_new_job
def clean_new_job(val):
    if val == '>4':
        return 5
    elif val == 'never':
        return 0
    else:
        try:
            return int(val)
        except:
            return None

all_data['last_new_job'] = all_data['last_new_job'].apply(clean_new_job)
all_data['last_new_job'] = all_data['last_new_job'].fillna(all_data['last_new_job'].median())

all_data.head()

enrollee_id      city  city_development_index   gender  \
0        25298  city_138                   0.836     Male   
1         4241  city_160                   0.920     Male   
2        24086   city_57                   0.866     Male   
3        26773   city_16                   0.910     Male   
4        32325  city_143                   0.740  Unknown   

       relevent_experience enrolled_university education_level  \
0   No relevent experience    Full time course     High School   
1   No relevent experience    Full time course     High School   
2   No relevent experience       no_enrollment        Graduate   
3  Has relevent experience       no_enrollment        Graduate   
4   No relevent experience    Full time course        Graduate   

  major_discipline  experience company_size company_type  last_new_job  \
0         No Major         5.0      100-500      Pvt Ltd           1.0   
1         No Major         5.0      Unknown      Unknown           1.0   
2             STEM        10.0      Unknown      Unknown           1.0   
3             STEM        21.0        50-99      Pvt Ltd           5.0   
4             STEM         5.0      Unknown      Unknown           0.0   

   training_hours  
0              45  
1              17  
2              50  
3             135  
4              17

In [159]:
# 확인
print(all_data.isnull().sum())

enrollee_id               0
city                      0
city_development_index    0
gender                    0
relevent_experience       0
enrolled_university       0
education_level           0
major_discipline          0
experience                0
company_size              0
company_type              0
last_new_job              0
training_hours            0
dtype: int64


In [160]:
import sklearn
print(sklearn.__all__)
import sklearn.model_selection
print(dir(sklearn.model_selection))
print(dir(sklearn.ensemble))

['calibration', 'cluster', 'covariance', 'cross_decomposition', 'datasets', 'decomposition', 'dummy', 'ensemble', 'exceptions', 'experimental', 'externals', 'feature_extraction', 'feature_selection', 'gaussian_process', 'inspection', 'isotonic', 'kernel_approximation', 'kernel_ridge', 'linear_model', 'manifold', 'metrics', 'mixture', 'model_selection', 'multiclass', 'multioutput', 'naive_bayes', 'neighbors', 'neural_network', 'pipeline', 'preprocessing', 'random_projection', 'semi_supervised', 'svm', 'tree', 'discriminant_analysis', 'impute', 'compose', 'clone', 'get_config', 'set_config', 'config_context', 'show_versions']
['BaseCrossValidator', 'BaseShuffleSplit', 'GridSearchCV', 'GroupKFold', 'GroupShuffleSplit', 'KFold', 'LearningCurveDisplay', 'LeaveOneGroupOut', 'LeaveOneOut', 'LeavePGroupsOut', 'LeavePOut', 'ParameterGrid', 'ParameterSampler', 'PredefinedSplit', 'RandomizedSearchCV', 'RepeatedKFold', 'RepeatedStratifiedKFold', 'ShuffleSplit', 'StratifiedGroupKFold', 'StratifiedK

In [161]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier

In [162]:
display(x_train.head())
display(y_train.head())

enrollee_id      city  city_development_index gender  \
0        25298  city_138                   0.836   Male   
1         4241  city_160                   0.920   Male   
2        24086   city_57                   0.866   Male   
3        26773   city_16                   0.910   Male   
4        32325  city_143                   0.740    NaN   

       relevent_experience enrolled_university education_level  \
0   No relevent experience    Full time course     High School   
1   No relevent experience    Full time course     High School   
2   No relevent experience       no_enrollment        Graduate   
3  Has relevent experience       no_enrollment        Graduate   
4   No relevent experience    Full time course        Graduate   

  major_discipline experience company_size company_type last_new_job  \
0              NaN          5      100-500      Pvt Ltd            1   
1              NaN          5          NaN          NaN            1   
2             STEM         10          NaN          NaN            1   
3             STEM        >20        50-99      Pvt Ltd           >4   
4             STEM          5          NaN          NaN        never   

   training_hours  
0              45  
1              17  
2              50  
3             135  
4              17

enrollee_id  target
0        25298     0.0
1         4241     1.0
2        24086     0.0
3        26773     0.0
4        32325     1.0

In [163]:
x_train = x_train.drop('enrollee_id', axis=1)
x_train.head(1)
x_test = x_test.drop('enrollee_id', axis=1)
x_test.head(1)
y_train = y_train.drop('enrollee_id', axis=1)['target']
y_train.head(1)

city  city_development_index gender     relevent_experience  \
0  city_138                   0.836   Male  No relevent experience   

  enrolled_university education_level major_discipline experience  \
0    Full time course     High School              NaN          5   

  company_size company_type last_new_job  training_hours  
0      100-500      Pvt Ltd            1              45

city  city_development_index gender     relevent_experience  \
0  city_23                   0.899   Male  No relevent experience   

  enrolled_university education_level major_discipline experience  \
0       no_enrollment        Graduate             STEM          2   

  company_size company_type last_new_job  training_hours  
0          NaN          NaN            1              23

0    0.0
Name: target, dtype: float64

In [164]:
x_train.shape
x_test.shape
y_train.shape

(12452, 12)

(6706, 12)

(12452,)

In [165]:
x_train_dummies = pd.get_dummies(x_train)
x_test_dummies = pd.get_dummies(x_test)

x_train_dummies.shape
x_test_dummies.shape

(12452, 185)

(6706, 181)

In [166]:
x_test_dummies = x_test_dummies.reindex(columns=x_train_dummies.columns, fill_value=0)

x_train_dummies.shape
x_test_dummies.shape

x_train_dummies.head(1)
x_test_dummies.head(1)

(12452, 185)

(6706, 185)

city_development_index  training_hours  city_city_1  city_city_10  \
0                   0.836              45        False         False   

   city_city_100  city_city_101  city_city_102  city_city_103  city_city_104  \
0          False          False          False          False          False   

   city_city_105  ...  company_type_NGO  company_type_Other  \
0          False  ...             False               False   

   company_type_Public Sector  company_type_Pvt Ltd  last_new_job_1  \
0                       False                  True            True   

   last_new_job_2  last_new_job_3  last_new_job_4  last_new_job_>4  \
0           False           False           False            False   

   last_new_job_never  
0               False  

[1 rows x 185 columns]

city_development_index  training_hours  city_city_1  city_city_10  \
0                   0.899              23        False         False   

   city_city_100  city_city_101  city_city_102  city_city_103  city_city_104  \
0          False          False          False          False          False   

   city_city_105  ...  company_type_NGO  company_type_Other  \
0          False  ...             False               False   

   company_type_Public Sector  company_type_Pvt Ltd  last_new_job_1  \
0                       False                 False            True   

   last_new_job_2  last_new_job_3  last_new_job_4  last_new_job_>4  \
0           False           False           False            False   

   last_new_job_never  
0               False  

[1 rows x 185 columns]

In [167]:
# 분류 - 타겟의 타입확인 : 정수로 변환
y_train.info()
y = y_train.astype(int)
y.dtypes

<class 'pandas.core.series.Series'>
RangeIndex: 12452 entries, 0 to 12451
Series name: target
Non-Null Count  Dtype  
--------------  -----  
12452 non-null  float64
dtypes: float64(1)
memory usage: 97.4 KB


dtype('int32')

In [168]:
# 분류 - 타겟의 비율 확인 : stratify = y 옵션의 여부
y.value_counts()

# 비율 3:1 -> 분류 문제는 일반적으로 statify 권장

target
0    9347
1    3105
Name: count, dtype: int64

In [169]:
X_train, X_val, y_train, y_val = train_test_split(x_train_dummies, y, test_size=0.3, random_state=2025, stratify=y)
# X_train, X_val, y_train, y_val = train_test_split(x_train_dummies, y, test_size=0.3, random_state=2025)
X_train.shape, X_val.shape, y_train.shape, y_val.shape

((8716, 185), (3736, 185), (8716,), (3736,))

In [170]:
y.value_counts(normalize=True)
y_train.value_counts(normalize=True)
y_val.value_counts(normalize=True)

target
0    0.750642
1    0.249358
Name: proportion, dtype: float64

target
0    0.750688
1    0.249312
Name: proportion, dtype: float64

target
0    0.750535
1    0.249465
Name: proportion, dtype: float64

In [171]:
print(x_train.dtypes)
print(X_train.dtypes)

city                       object
city_development_index    float64
gender                     object
relevent_experience        object
enrolled_university        object
education_level            object
major_discipline           object
experience                 object
company_size               object
company_type               object
last_new_job               object
training_hours              int64
dtype: object
city_development_index    float64
training_hours              int64
city_city_1                  bool
city_city_10                 bool
city_city_100                bool
                           ...   
last_new_job_2               bool
last_new_job_3               bool
last_new_job_4               bool
last_new_job_>4              bool
last_new_job_never           bool
Length: 185, dtype: object


In [172]:
X_train.select_dtypes(bool).sum()

city_city_1              7
city_city_10            40
city_city_100          136
city_city_101           38
city_city_102          133
                      ... 
last_new_job_2        1358
last_new_job_3         494
last_new_job_4         481
last_new_job_>4       1520
last_new_job_never    1142
Length: 183, dtype: int64

In [173]:
for dtype in X_train.dtypes.unique():
    print(f"\n🔹 dtype: {dtype}\n")
    cols = X_train.select_dtypes(include=[dtype]).columns
    for col in cols:
        print(f"Column: {col}")
        print(X_train[col].value_counts())
        print('-' * 40)



🔹 dtype: float64

Column: city_development_index
city_development_index
0.920    2395
0.624    1196
0.910     702
0.926     590
0.698     314
         ... 
0.824       2
0.807       2
0.781       1
0.625       1
0.649       1
Name: count, Length: 92, dtype: int64
----------------------------------------

🔹 dtype: int64

Column: training_hours
training_hours
28     150
20     136
50     134
17     134
18     132
      ... 
234      2
244      2
310      2
272      1
238      1
Name: count, Length: 241, dtype: int64
----------------------------------------

🔹 dtype: bool

Column: city_city_1
city_city_1
False    8709
True        7
Name: count, dtype: int64
----------------------------------------
Column: city_city_10
city_city_10
False    8676
True       40
Name: count, dtype: int64
----------------------------------------
Column: city_city_100
city_city_100
False    8580
True      136
Name: count, dtype: int64
----------------------------------------
Column: city_city_101
city_city_101

company_size_10000+
False    7773
True      943
Name: count, dtype: int64
----------------------------------------
Column: company_size_50-99
company_size_50-99
False    7312
True     1404
Name: count, dtype: int64
----------------------------------------
Column: company_size_500-999
company_size_500-999
False    8301
True      415
Name: count, dtype: int64
----------------------------------------
Column: company_size_5000-9999
company_size_5000-9999
False    8462
True      254
Name: count, dtype: int64
----------------------------------------
Column: company_size_<10
company_size_<10
False    8135
True      581
Name: count, dtype: int64
----------------------------------------
Column: company_type_Early Stage Startup
company_type_Early Stage Startup
False    8451
True      265
Name: count, dtype: int64
----------------------------------------
Column: company_type_Funded Startup
company_type_Funded Startup
False    8265
True      451
Name: count, dtype: int64
--------------------------

In [181]:
X_train.shape
y_train.shape

(8716, 185)

(8716,)

In [194]:
model = rf.fit(X_train,y_train)
model

RandomForestClassifier()

In [183]:
import sklearn.metrics
print(dir(sklearn.metrics))

['ConfusionMatrixDisplay', 'DetCurveDisplay', 'DistanceMetric', 'PrecisionRecallDisplay', 'PredictionErrorDisplay', 'RocCurveDisplay', '__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '_base', '_classification', '_dist_metrics', '_pairwise_distances_reduction', '_pairwise_fast', '_plot', '_ranking', '_regression', '_scorer', 'accuracy_score', 'adjusted_mutual_info_score', 'adjusted_rand_score', 'auc', 'average_precision_score', 'balanced_accuracy_score', 'brier_score_loss', 'calinski_harabasz_score', 'check_scoring', 'class_likelihood_ratios', 'classification_report', 'cluster', 'cohen_kappa_score', 'completeness_score', 'confusion_matrix', 'consensus_score', 'coverage_error', 'd2_absolute_error_score', 'd2_pinball_score', 'd2_tweedie_score', 'davies_bouldin_score', 'dcg_score', 'det_curve', 'euclidean_distances', 'explained_variance_score', 'f1_score', 'fbeta_score', 'fowlkes_mallows_score', 'get_scorer', 

In [184]:
from sklearn.metrics import accuracy_score, f1_score, recall_score, roc_auc_score, precision_score 

In [185]:
# 에측
predict_train_label = rf.predict(x_test_dummies)
predict_train_label
predict_train_proba = rf.predict(x_test_dummies)
predict_train_proba

array([1, 1, 0, ..., 0, 0, 1])

array([1, 1, 0, ..., 0, 0, 1])

In [201]:
y_pred = rf.predict(X_val) # 예측된 클래스 라벨
y_pred
y_pred_proba_1 = rf.predict_proba(X_val)[:,1] # 클래스 1 확률
y_pred_proba_1
y_pred_proba_0 = rf.predict_proba(X_val)[:,0] # 클래스 0확률
y_pred_proba_0

array([0, 0, 0, ..., 0, 0, 0])

array([0.49, 0.31, 0.01, ..., 0.11, 0.02, 0.17])

array([0.51, 0.69, 0.99, ..., 0.89, 0.98, 0.83])

In [205]:
y_pred_proba = rf.predict_proba(X_val)[:, 1]  

print('Accuracy:', accuracy_score(y_val, y_pred))
print('F1 score:', f1_score(y_val, y_pred))
print('Recall:', recall_score(y_val, y_pred))
print('Precision:', precision_score(y_val, y_pred))
print('ROC AUC:', roc_auc_score(y_val, y_pred_proba))


Accuracy: 0.7813169164882227
F1 score: 0.49033063006862126
Recall: 0.4216738197424893
Precision: 0.5856929955290611
ROC AUC: 0.7747490938757018


# 기출 8회 변형


# 유형 1

### 1-1) 대륙별 맥주 평균 소비량을 구하라. 5번째로 소비가 많은 대륙의 2번째로 맥주 소비량이 많은 국가의 맥주 소비량을 구하여라 

DataUrl = https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p1_1.csv  
데이터 출처 : kaggle  
  
데이터 설명 : 대륙,국가별 맥주 소비량  

In [209]:
import pandas as pd
df_o = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p1_1.csv')
df = df_o.copy()
df.head()

대륙       국가명  소비량
0  Africa   Algeria   25
1  Africa    Angola  217
2  Africa     Benin   34
3  Africa  Botswana  173
4  Africa   Burundi   88

In [214]:
df.info()
df.nunique()
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 177 entries, 0 to 176
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   대륙      177 non-null    object
 1   국가명     177 non-null    object
 2   소비량     177 non-null    int64 
dtypes: int64(1), object(2)
memory usage: 4.3+ KB


대륙       6
국가명    177
소비량    122
dtype: int64

대륙     0
국가명    0
소비량    0
dtype: int64

In [255]:
df[df['대륙'] == df.groupby('대륙', as_index=False)['소비량'].sum(numeric_only=True).sort_values(by='소비량', ascending=False).iloc[4]['대륙']].groupby('국가명', as_index=False)['소비량'].sum().sort_values(by='소비량', ascending=False).iloc[1]['소비량']

140

### 1-2) 총 방문객 수는 관광,방문,유햑연수,취업,기타 방문자 숫자의 합으로 정의한다. 각 국가별 관광 / 총 방문객 수의 비율을 구하고 이를 관광 비율이라 한다. 관광비율이 2번째로 높은 국가의 취업으로 출국한 인원과 관광 숫자가 3번째로 높은 국가의 방문으로 출국한 인원의 합을 구하여라  

DataUrl = https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p1_2.csv   
데이터 출처 : https://know.tour.go.kr/stat/exitTourStatDis19Re.do 
  
후처리 데이터 설명 : 목적별 행선지별 출국  

In [257]:
import pandas as pd
df_o = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p1_2.csv')
df= df_o.copy()
df.head()

국가       관광      방문    유학연수    취업     기타
0    뉴질랜드  24724.0  6320.0  4911.0  56.0  637.0
1    피  지   3889.0   564.0    98.0  15.0   40.0
2     벨기에    243.0    58.0    30.0   7.0   25.0
3     멕시코   1069.0   335.0   103.0  32.0   48.0
4  우즈베키스탄   3994.0  1456.0    97.0   4.0  137.0

In [259]:
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64 entries, 0 to 63
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   국가      64 non-null     object 
 1   관광      64 non-null     float64
 2   방문      64 non-null     float64
 3   유학연수    64 non-null     float64
 4   취업      64 non-null     float64
 5   기타      64 non-null     float64
dtypes: float64(5), object(1)
memory usage: 3.1+ KB


관광             방문          유학연수           취업            기타
count      64.000000      64.000000     64.000000    64.000000     64.000000
mean    54221.921875    8802.578125   3612.437500   174.828125   1161.281250
std    151948.765486   25648.424098  10424.900911   425.183865   3007.993512
min        23.000000       6.000000      0.000000     0.000000      5.000000
25%       451.500000     110.750000     29.750000     7.000000     44.250000
50%      3941.500000     707.500000    136.500000    30.500000    138.000000
75%     35786.250000    5599.250000   1338.250000   126.250000    920.500000
max    902723.000000  152675.000000  52129.000000  2497.000000  18054.000000

In [263]:
# dir(df.describe())

In [269]:
df['관광비율'] = df['관광']/df['방문']
df_n2 = df.sort_values(by='관광비율', ascending=False).iloc[1]
df_n2
df_n3 = df.sort_values(by='관광', ascending=False).iloc[2]
df_n3

result = df_n2['취업'] + df_n3['방문']
result

국가            마카오
관광        24237.0
방문         1603.0
유학연수         81.0
취업            4.0
기타          123.0
관광비율    15.119775
Name: 36, dtype: object

국가           태  국
관광       416800.0
방문        30813.0
유학연수       4133.0
취업          197.0
기타         3111.0
관광비율    13.526758
Name: 29, dtype: object

30817.0

In [267]:
df.nunique() # 식별자로 활용할 국가의 중복 확인

국가      64
관광      64
방문      63
유학연수    60
취업      49
기타      62
관광비율    64
dtype: int64

### 1-3) CO(GT),NMHC(GT) 두 컬럼을 min-max normalize를 진행하고 각 컬럼의 모표준편차의 차이의 절댓값을 소숫점 이하 3째 자리까지 구하여라  
  
 

In [272]:
import pandas as pd
df_o = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p1_3.csv')
df = df_o.copy()
df.head(5)

Date      Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)
0  10/03/2004  18.00.00     2.6       1360.0     150.0      11.9
1  10/03/2004  19.00.00     2.0       1292.0     112.0       9.4
2  10/03/2004  20.00.00     2.2       1402.0      88.0       9.0
3  10/03/2004  21.00.00     2.2       1376.0      80.0       9.2
4  10/03/2004  22.00.00     1.6       1272.0      51.0       6.5

In [273]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9357 entries, 0 to 9356
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Date         9357 non-null   object 
 1   Time         9357 non-null   object 
 2   CO(GT)       9357 non-null   float64
 3   PT08.S1(CO)  9357 non-null   float64
 4   NMHC(GT)     9357 non-null   float64
 5   C6H6(GT)     9357 non-null   float64
dtypes: float64(4), object(2)
memory usage: 438.7+ KB


In [283]:
import numpy as np

def min_max_nor(x):
    nmn = (x - x.mean()) / (x.max() - x.min())
    return nmn


df['CO(GT)'] = df['CO(GT)'].agg(min_max_nor)
df['NMHC(GT)'] = df['NMHC(GT)'].agg(min_max_nor)
# df

co_std = df['CO(GT)'].std(ddof=0)
nmhc_std = df['NMHC(GT)'].std(ddof=0)

round(abs(co_std - nmhc_std), 3)


0.266

In [284]:
import numpy as np

def min_max_nor(x):
    nmn = (x - x.mean()) / (x.max() - x.min())
    return nmn


df['CO(GT)'] = df['CO(GT)'].agg(min_max_nor)
df['NMHC(GT)'] = df['NMHC(GT)'].agg(min_max_nor)
# df

def pop_std (arr):
    mean = arr.mean()
    var = sum((x - mean)**2 for x in arr) / len(arr)
    return var ** 0.5

round(abs(pop_std(df['CO(GT)']) - pop_std(df['NMHC(GT)'])),3)

0.266

# 유형 2

날씨에 따른 지하철 이용자수   
train = https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p2_train.csv   
test = https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p2_test.csv  
  
종속변수 : Y
평가지표 : mae  

In [365]:
import pandas as pd
train = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p2_train2.csv')
test = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p2_test2.csv')

display(train.head(2))
test.head(2)

ID   C1   C2   C3   C4        Date  Weekday  Temperature   Humidity  \
0  ID_1300  X_2  Y_0  Z_3  W_2  2023-03-27        0    10.533315  88.101605   
1  ID_1301  X_0  Y_3  Z_2  W_9  2023-10-12        3    -1.698046  42.811039   

   Wind_Speed  Precipitation     Pressure  Visibility   Y  
0   13.632211       7.814899  1046.992484    6.100292  27  
1    6.165366       1.139649   985.766276    5.720027  -5

ID   C1   C2   C3    C4        Date  Weekday  Temperature   Humidity  \
0  ID_4780  X_1  Y_2  Z_2  W_11  2023-07-27        3    28.460744  32.869427   
1   ID_217  X_2  Y_2  Z_0   W_9  2023-09-05        1    33.699748  20.366573   

   Wind_Speed  Precipitation     Pressure  Visibility  
0   13.978067       1.453519   980.152660    3.091170  
1    9.032971       1.430375  1014.322511    7.788306

In [366]:
train.info()
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3934 entries, 0 to 3933
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ID             3934 non-null   object 
 1   C1             3934 non-null   object 
 2   C2             3934 non-null   object 
 3   C3             3934 non-null   object 
 4   C4             3934 non-null   object 
 5   Date           3934 non-null   object 
 6   Weekday        3934 non-null   int64  
 7   Temperature    3934 non-null   float64
 8   Humidity       3934 non-null   float64
 9   Wind_Speed     3934 non-null   float64
 10  Precipitation  3934 non-null   float64
 11  Pressure       3934 non-null   float64
 12  Visibility     3934 non-null   float64
 13  Y              3934 non-null   int64  
dtypes: float64(6), int64(2), object(6)
memory usage: 430.4+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1300 entries, 0 to 1299
Data columns (total 13 columns):
 #   Column       

In [367]:
y = train['Y']
y.shape
train = train.drop('Y', axis=1)
train.shape
test.shape

(3934,)

(3934, 13)

(1300, 13)

In [376]:
# y값이 1개 이하인 클래스 제거
value_counts = y.value_counts()
value_counts
valid_classes = value_counts[value_counts >= 2].index
valid_classes

# 해당 클래스만 필터링
mask = y.isin(valid_classes)
mask
train = train[mask]
y = y[mask]
train.shape
test.shape
y.shape


Y
 40     53
 46     49
 29     49
 42     47
 43     47
        ..
 109     2
-38      2
 127     2
-50      2
-46      2
Name: count, Length: 174, dtype: int64

Index([ 40,  46,  29,  42,  43,  53,  44,  30,  48,  39,
       ...
       119, 126, 120, -57, 133, 109, -38, 127, -50, -46],
      dtype='int64', name='Y', length=174)

0       True
1       True
2       True
3       True
4       True
        ... 
3929    True
3930    True
3931    True
3932    True
3933    True
Name: Y, Length: 3919, dtype: bool

(3919, 13)

(1297, 13)

(3919,)

In [298]:
total = pd.concat([train, test], ignore_index=True)
total

ID   C1   C2   C3   C4        Date  Weekday  Temperature  \
0     ID_1300  X_2  Y_0  Z_3  W_2  2023-03-27        0    10.533315   
1     ID_1301  X_0  Y_3  Z_2  W_9  2023-10-12        3    -1.698046   
2     ID_1302  X_0  Y_4  Z_3  W_5  2023-03-01        2     6.705572   
3     ID_1303  X_2  Y_4  Z_6  W_5  2023-02-23        3    13.500946   
4     ID_1304  X_2  Y_3  Z_2  W_3  2023-04-03        0    -7.600114   
...       ...  ...  ...  ...  ...         ...      ...          ...   
5229  ID_2462  X_2  Y_3  Z_5  W_1  2023-03-04        5     5.695576   
5230  ID_5222  X_1  Y_4  Z_2  W_6  2023-06-11        6    -8.581502   
5231  ID_4339  X_0  Y_1  Z_1  W_6  2023-04-17        0    -3.799010   
5232  ID_2831  X_0  Y_4  Z_5  W_7  2023-07-05        2    24.059002   
5233   ID_683  X_0  Y_2  Z_4  W_7  2023-11-12        6    -4.997380   

       Humidity  Wind_Speed  Precipitation     Pressure  Visibility  
0     88.101605   13.632211       7.814899  1046.992484    6.100292  
1     42.811039    6.165366       1.139649   985.766276    5.720027  
2     30.026276   12.749443       2.134840   961.231073    1.815005  
3     70.672815    3.738548       5.520962   972.253733    9.343229  
4     33.941831    4.697525       4.399400  1017.066735    6.330298  
...         ...         ...            ...          ...         ...  
5229  68.823684    7.065409       6.508675  1007.692743    1.152456  
5230  20.714936    8.487859      15.685114  1005.370528    7.589226  
5231  82.193000   10.877006      13.334141   988.465067    3.089898  
5232  53.132618   10.268096       3.516653   980.608151    7.851758  
5233  23.182071   11.041030       8.222889   995.488966    5.547096  

[5234 rows x 13 columns]

In [325]:
total_num = total.select_dtypes(exclude='object')
total_num

Weekday  Temperature   Humidity  Wind_Speed  Precipitation     Pressure  \
0           0    10.533315  88.101605   13.632211       7.814899  1046.992484   
1           3    -1.698046  42.811039    6.165366       1.139649   985.766276   
2           2     6.705572  30.026276   12.749443       2.134840   961.231073   
3           3    13.500946  70.672815    3.738548       5.520962   972.253733   
4           0    -7.600114  33.941831    4.697525       4.399400  1017.066735   
...       ...          ...        ...         ...            ...          ...   
5229        5     5.695576  68.823684    7.065409       6.508675  1007.692743   
5230        6    -8.581502  20.714936    8.487859      15.685114  1005.370528   
5231        0    -3.799010  82.193000   10.877006      13.334141   988.465067   
5232        2    24.059002  53.132618   10.268096       3.516653   980.608151   
5233        6    -4.997380  23.182071   11.041030       8.222889   995.488966   

      Visibility  
0       6.100292  
1       5.720027  
2       1.815005  
3       9.343229  
4       6.330298  
...          ...  
5229    1.152456  
5230    7.589226  
5231    3.089898  
5232    7.851758  
5233    5.547096  

[5234 rows x 7 columns]

In [305]:
import sklearn
sklearn.__all__

['calibration',
 'cluster',
 'covariance',
 'cross_decomposition',
 'datasets',
 'decomposition',
 'dummy',
 'ensemble',
 'exceptions',
 'experimental',
 'externals',
 'feature_extraction',
 'feature_selection',
 'gaussian_process',
 'inspection',
 'isotonic',
 'kernel_approximation',
 'kernel_ridge',
 'linear_model',
 'manifold',
 'metrics',
 'mixture',
 'model_selection',
 'multiclass',
 'multioutput',
 'naive_bayes',
 'neighbors',
 'neural_network',
 'pipeline',
 'preprocessing',
 'random_projection',
 'semi_supervised',
 'svm',
 'tree',
 'discriminant_analysis',
 'impute',
 'compose',
 'clone',
 'get_config',
 'set_config',
 'config_context',
 'show_versions']

In [308]:
dir(sklearn.preprocessing)

['Binarizer',
 'FunctionTransformer',
 'KBinsDiscretizer',
 'KernelCenterer',
 'LabelBinarizer',
 'LabelEncoder',
 'MaxAbsScaler',
 'MinMaxScaler',
 'MultiLabelBinarizer',
 'Normalizer',
 'OneHotEncoder',
 'OrdinalEncoder',
 'PolynomialFeatures',
 'PowerTransformer',
 'QuantileTransformer',
 'RobustScaler',
 'SplineTransformer',
 'StandardScaler',
 'TargetEncoder',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '_csr_polynomial_expansion',
 '_data',
 '_discretization',
 '_encoders',
 '_function_transformer',
 '_label',
 '_polynomial',
 '_target_encoder',
 '_target_encoder_fast',
 'add_dummy_feature',
 'binarize',
 'label_binarize',
 'maxabs_scale',
 'minmax_scale',
 'normalize',
 'power_transform',
 'quantile_transform',
 'robust_scale',
 'scale']

# 숫자형 컬럼 스케일링

In [316]:
from sklearn.preprocessing import MinMaxScaler
mms = MinMaxScaler()

In [326]:
total[total_num.columns]

Weekday  Temperature   Humidity  Wind_Speed  Precipitation     Pressure  \
0           0    10.533315  88.101605   13.632211       7.814899  1046.992484   
1           3    -1.698046  42.811039    6.165366       1.139649   985.766276   
2           2     6.705572  30.026276   12.749443       2.134840   961.231073   
3           3    13.500946  70.672815    3.738548       5.520962   972.253733   
4           0    -7.600114  33.941831    4.697525       4.399400  1017.066735   
...       ...          ...        ...         ...            ...          ...   
5229        5     5.695576  68.823684    7.065409       6.508675  1007.692743   
5230        6    -8.581502  20.714936    8.487859      15.685114  1005.370528   
5231        0    -3.799010  82.193000   10.877006      13.334141   988.465067   
5232        2    24.059002  53.132618   10.268096       3.516653   980.608151   
5233        6    -4.997380  23.182071   11.041030       8.222889   995.488966   

      Visibility  
0       6.100292  
1       5.720027  
2       1.815005  
3       9.343229  
4       6.330298  
...          ...  
5229    1.152456  
5230    7.589226  
5231    3.089898  
5232    7.851758  
5233    5.547096  

[5234 rows x 7 columns]

In [329]:
total_mms = mms.fit_transform(total[total_num.columns])
total_mms

array([[0.        , 0.4563149 , 0.97656474, ..., 0.39092117, 0.96995136,
        0.56667808],
       [0.5       , 0.1844669 , 0.41019783, ..., 0.05685423, 0.35766197,
        0.52442025],
       [0.33333333, 0.37124144, 0.250322  , ..., 0.10665918, 0.112299  ,
        0.09046477],
       ...,
       [0.        , 0.13777192, 0.90267654, ..., 0.66713499, 0.38465109,
        0.2321405 ],
       [0.33333333, 0.75692996, 0.53927109, ..., 0.17581286, 0.30607842,
        0.76131419],
       [1.        , 0.11113756, 0.16473396, ..., 0.41133929, 0.45489321,
        0.50520279]])

In [335]:
y.value_counts()

Y
 40     53
 29     49
 46     49
 42     47
 43     47
        ..
 125     1
-52      1
 137     1
 116     1
 142     1
Name: count, Length: 189, dtype: int64

In [378]:
total_mms_df = pd.DataFrame(total_mms, columns=total_num.columns, index=total.index)
total_mms_df

total[total_num.columns] = total_mms_df[total_num.columns]
total

Weekday  Temperature  Humidity  Wind_Speed  Precipitation  Pressure  \
0     0.000000     0.456315  0.976565    0.909051       0.390921  0.969951   
1     0.500000     0.184467  0.410198    0.411095       0.056854  0.357662   
2     0.333333     0.371241  0.250322    0.850181       0.106659  0.112299   
3     0.500000     0.522272  0.758614    0.249253       0.276120  0.222531   
4     0.000000     0.053291  0.299287    0.313206       0.219990  0.670681   
...        ...          ...       ...         ...            ...       ...   
5229  0.833333     0.348794  0.735491    0.471118       0.325550  0.576936   
5230  1.000000     0.031479  0.133882    0.565980       0.784791  0.553713   
5231  0.000000     0.137772  0.902677    0.725309       0.667135  0.384651   
5232  0.333333     0.756930  0.539271    0.684702       0.175813  0.306078   
5233  1.000000     0.111138  0.164734    0.736248       0.411339  0.454893   

      Visibility  
0       0.566678  
1       0.524420  
2       0.090465  
3       0.927058  
4       0.592238  
...          ...  
5229    0.016837  
5230    0.732140  
5231    0.232140  
5232    0.761314  
5233    0.505203  

[5234 rows x 7 columns]

ID   C1   C2   C3   C4        Date   Weekday  Temperature  \
0     ID_1300  X_2  Y_0  Z_3  W_2  2023-03-27  0.000000     0.456315   
1     ID_1301  X_0  Y_3  Z_2  W_9  2023-10-12  0.500000     0.184467   
2     ID_1302  X_0  Y_4  Z_3  W_5  2023-03-01  0.333333     0.371241   
3     ID_1303  X_2  Y_4  Z_6  W_5  2023-02-23  0.500000     0.522272   
4     ID_1304  X_2  Y_3  Z_2  W_3  2023-04-03  0.000000     0.053291   
...       ...  ...  ...  ...  ...         ...       ...          ...   
5229  ID_2462  X_2  Y_3  Z_5  W_1  2023-03-04  0.833333     0.348794   
5230  ID_5222  X_1  Y_4  Z_2  W_6  2023-06-11  1.000000     0.031479   
5231  ID_4339  X_0  Y_1  Z_1  W_6  2023-04-17  0.000000     0.137772   
5232  ID_2831  X_0  Y_4  Z_5  W_7  2023-07-05  0.333333     0.756930   
5233   ID_683  X_0  Y_2  Z_4  W_7  2023-11-12  1.000000     0.111138   

      Humidity  Wind_Speed  Precipitation  Pressure  Visibility  
0     0.976565    0.909051       0.390921  0.969951    0.566678  
1     0.410198    0.411095       0.056854  0.357662    0.524420  
2     0.250322    0.850181       0.106659  0.112299    0.090465  
3     0.758614    0.249253       0.276120  0.222531    0.927058  
4     0.299287    0.313206       0.219990  0.670681    0.592238  
...        ...         ...            ...       ...         ...  
5229  0.735491    0.471118       0.325550  0.576936    0.016837  
5230  0.133882    0.565980       0.784791  0.553713    0.732140  
5231  0.902677    0.725309       0.667135  0.384651    0.232140  
5232  0.539271    0.684702       0.175813  0.306078    0.761314  
5233  0.164734    0.736248       0.411339  0.454893    0.505203  

[5234 rows x 13 columns]

# 날짜 -> 문자열에서 날짜타입으로 변경

In [380]:
import datetime

In [382]:
datetime.__all__

('date',
 'datetime',
 'time',
 'timedelta',
 'timezone',
 'tzinfo',
 'MINYEAR',
 'MAXYEAR',
 'UTC')

In [386]:
dir(datetime.date)

['__add__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__ne__',
 '__new__',
 '__radd__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rsub__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 'ctime',
 'day',
 'fromisocalendar',
 'fromisoformat',
 'fromordinal',
 'fromtimestamp',
 'isocalendar',
 'isoformat',
 'isoweekday',
 'max',
 'min',
 'month',
 'replace',
 'resolution',
 'strftime',
 'timetuple',
 'today',
 'toordinal',
 'weekday',
 'year']

In [404]:
total['Date']

0       2023-03-27
1       2023-10-12
2       2023-03-01
3       2023-02-23
4       2023-04-03
           ...    
5229    2023-03-04
5230    2023-06-11
5231    2023-04-17
5232    2023-07-05
5233    2023-11-12
Name: Date, Length: 5234, dtype: object

In [409]:
import pandas as pd

total['Date'] = pd.to_datetime(total['Date'], format='%Y-%m-%d')
total.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5234 entries, 0 to 5233
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   ID             5234 non-null   object        
 1   C1             5234 non-null   object        
 2   C2             5234 non-null   object        
 3   C3             5234 non-null   object        
 4   C4             5234 non-null   object        
 5   Date           5234 non-null   datetime64[ns]
 6   Weekday        5234 non-null   float64       
 7   Temperature    5234 non-null   float64       
 8   Humidity       5234 non-null   float64       
 9   Wind_Speed     5234 non-null   float64       
 10  Precipitation  5234 non-null   float64       
 11  Pressure       5234 non-null   float64       
 12  Visibility     5234 non-null   float64       
dtypes: datetime64[ns](1), float64(7), object(5)
memory usage: 531.7+ KB


# 범주형 원핫인코딩

In [412]:
total_obj = total.select_dtypes(include='object')
total_obj

ID   C1   C2   C3   C4
0     ID_1300  X_2  Y_0  Z_3  W_2
1     ID_1301  X_0  Y_3  Z_2  W_9
2     ID_1302  X_0  Y_4  Z_3  W_5
3     ID_1303  X_2  Y_4  Z_6  W_5
4     ID_1304  X_2  Y_3  Z_2  W_3
...       ...  ...  ...  ...  ...
5229  ID_2462  X_2  Y_3  Z_5  W_1
5230  ID_5222  X_1  Y_4  Z_2  W_6
5231  ID_4339  X_0  Y_1  Z_1  W_6
5232  ID_2831  X_0  Y_4  Z_5  W_7
5233   ID_683  X_0  Y_2  Z_4  W_7

[5234 rows x 5 columns]

In [414]:
total_dm = pd.get_dummies(total_obj)
total_dm

ID_ID_100  ID_ID_1000  ID_ID_1003  ID_ID_1007  ID_ID_1010  ID_ID_1012  \
0         False       False       False       False       False       False   
1         False       False       False       False       False       False   
2         False       False       False       False       False       False   
3         False       False       False       False       False       False   
4         False       False       False       False       False       False   
...         ...         ...         ...         ...         ...         ...   
5229      False       False       False       False       False       False   
5230      False       False       False       False       False       False   
5231      False       False       False       False       False       False   
5232      False       False       False       False       False       False   
5233      False       False       False       False       False       False   

      ID_ID_1020  ID_ID_1026  ID_ID_1027  ID_ID_1032  ...  C4_W_10  C4_W_11  \
0          False       False       False       False  ...    False    False   
1          False       False       False       False  ...    False    False   
2          False       False       False       False  ...    False    False   
3          False       False       False       False  ...    False    False   
4          False       False       False       False  ...    False    False   
...          ...         ...         ...         ...  ...      ...      ...   
5229       False       False       False       False  ...    False    False   
5230       False       False       False       False  ...    False    False   
5231       False       False       False       False  ...    False    False   
5232       False       False       False       False  ...    False    False   
5233       False       False       False       False  ...    False    False   

      C4_W_2  C4_W_3  C4_W_4  C4_W_5  C4_W_6  C4_W_7  C4_W_8  C4_W_9  
0       True   False   False   False   False   False   False   False  
1      False   False   False   False   False   False   False    True  
2      False   False   False    True   False   False   False   False  
3      False   False   False    True   False   False   False   False  
4      False    True   False   False   False   False   False   False  
...      ...     ...     ...     ...     ...     ...     ...     ...  
5229   False   False   False   False   False   False   False   False  
5230   False   False   False   False    True   False   False   False  
5231   False   False   False   False    True   False   False   False  
5232   False   False   False   False   False    True   False   False  
5233   False   False   False   False   False    True   False   False  

[5234 rows x 4266 columns]

In [416]:
total = total.drop(columns=total_dm.columns, errors='ignore')  # 기존 더미 컬럼 제거
total = pd.concat([total, total_dm], axis=1)  # 한 번에 합치기
total

ID   C1   C2   C3   C4       Date   Weekday  Temperature  Humidity  \
0     ID_1300  X_2  Y_0  Z_3  W_2 2023-03-27  0.000000     0.456315  0.976565   
1     ID_1301  X_0  Y_3  Z_2  W_9 2023-10-12  0.500000     0.184467  0.410198   
2     ID_1302  X_0  Y_4  Z_3  W_5 2023-03-01  0.333333     0.371241  0.250322   
3     ID_1303  X_2  Y_4  Z_6  W_5 2023-02-23  0.500000     0.522272  0.758614   
4     ID_1304  X_2  Y_3  Z_2  W_3 2023-04-03  0.000000     0.053291  0.299287   
...       ...  ...  ...  ...  ...        ...       ...          ...       ...   
5229  ID_2462  X_2  Y_3  Z_5  W_1 2023-03-04  0.833333     0.348794  0.735491   
5230  ID_5222  X_1  Y_4  Z_2  W_6 2023-06-11  1.000000     0.031479  0.133882   
5231  ID_4339  X_0  Y_1  Z_1  W_6 2023-04-17  0.000000     0.137772  0.902677   
5232  ID_2831  X_0  Y_4  Z_5  W_7 2023-07-05  0.333333     0.756930  0.539271   
5233   ID_683  X_0  Y_2  Z_4  W_7 2023-11-12  1.000000     0.111138  0.164734   

      Wind_Speed  ...  C4_W_10  C4_W_11  C4_W_2  C4_W_3  C4_W_4  C4_W_5  \
0       0.909051  ...    False    False    True   False   False   False   
1       0.411095  ...    False    False   False   False   False   False   
2       0.850181  ...    False    False   False   False   False    True   
3       0.249253  ...    False    False   False   False   False    True   
4       0.313206  ...    False    False   False    True   False   False   
...          ...  ...      ...      ...     ...     ...     ...     ...   
5229    0.471118  ...    False    False   False   False   False   False   
5230    0.565980  ...    False    False   False   False   False   False   
5231    0.725309  ...    False    False   False   False   False   False   
5232    0.684702  ...    False    False   False   False   False   False   
5233    0.736248  ...    False    False   False   False   False   False   

      C4_W_6  C4_W_7  C4_W_8  C4_W_9  
0      False   False   False   False  
1      False   False   False    True  
2      False   False   False   False  
3      False   False   False   False  
4      False   False   False   False  
...      ...     ...     ...     ...  
5229   False   False   False   False  
5230    True   False   False   False  
5231    True   False   False   False  
5232   False    True   False   False  
5233   False    True   False   False  

[5234 rows x 4279 columns]

In [417]:
train = total.iloc[:len(train)]
train
test = total.iloc[len(train):]
test

ID   C1   C2   C3   C4       Date   Weekday  Temperature  Humidity  \
0     ID_1300  X_2  Y_0  Z_3  W_2 2023-03-27  0.000000     0.456315  0.976565   
1     ID_1301  X_0  Y_3  Z_2  W_9 2023-10-12  0.500000     0.184467  0.410198   
2     ID_1302  X_0  Y_4  Z_3  W_5 2023-03-01  0.333333     0.371241  0.250322   
3     ID_1303  X_2  Y_4  Z_6  W_5 2023-02-23  0.500000     0.522272  0.758614   
4     ID_1304  X_2  Y_3  Z_2  W_3 2023-04-03  0.000000     0.053291  0.299287   
...       ...  ...  ...  ...  ...        ...       ...          ...       ...   
3914  ID_5214  X_1  Y_3  Z_6  W_3 2023-10-06  0.666667     0.412062  0.974199   
3915  ID_5215  X_1  Y_3  Z_1  W_2 2023-08-12  0.833333     0.835312  0.380156   
3916  ID_5216  X_0  Y_4  Z_5  W_7 2023-08-06  1.000000     0.513420  0.064389   
3917  ID_5217  X_2  Y_4  Z_0  W_6 2023-08-25  0.666667     0.731126  0.988271   
3918  ID_5218  X_2  Y_4  Z_5  W_3 2023-07-23  1.000000     0.018962  0.623955   

      Wind_Speed  ...  C4_W_10  C4_W_11  C4_W_2  C4_W_3  C4_W_4  C4_W_5  \
0       0.909051  ...    False    False    True   False   False   False   
1       0.411095  ...    False    False   False   False   False   False   
2       0.850181  ...    False    False   False   False   False    True   
3       0.249253  ...    False    False   False   False   False    True   
4       0.313206  ...    False    False   False    True   False   False   
...          ...  ...      ...      ...     ...     ...     ...     ...   
3914    0.327240  ...    False    False   False    True   False   False   
3915    0.507084  ...    False    False    True   False   False   False   
3916    0.346705  ...    False    False   False   False   False   False   
3917    0.994474  ...    False    False   False   False   False   False   
3918    0.357530  ...    False    False   False    True   False   False   

      C4_W_6  C4_W_7  C4_W_8  C4_W_9  
0      False   False   False   False  
1      False   False   False    True  
2      False   False   False   False  
3      False   False   False   False  
4      False   False   False   False  
...      ...     ...     ...     ...  
3914   False   False   False   False  
3915   False   False   False   False  
3916   False    True   False   False  
3917    True   False   False   False  
3918   False   False   False   False  

[3919 rows x 4279 columns]

ID   C1   C2   C3    C4       Date   Weekday  Temperature  \
3919  ID_5219  X_2  Y_0  Z_6   W_3 2023-12-05  0.166667     0.464513   
3920  ID_5220  X_2  Y_2  Z_6   W_0 2023-01-19  0.500000     0.901402   
3921  ID_5221  X_1  Y_4  Z_0  W_10 2023-06-29  0.500000     0.224299   
3922  ID_5222  X_1  Y_4  Z_2   W_6 2023-06-11  1.000000     0.031479   
3923  ID_5223  X_1  Y_0  Z_4   W_0 2023-10-12  0.500000     0.433424   
...       ...  ...  ...  ...   ...        ...       ...          ...   
5229  ID_2462  X_2  Y_3  Z_5   W_1 2023-03-04  0.833333     0.348794   
5230  ID_5222  X_1  Y_4  Z_2   W_6 2023-06-11  1.000000     0.031479   
5231  ID_4339  X_0  Y_1  Z_1   W_6 2023-04-17  0.000000     0.137772   
5232  ID_2831  X_0  Y_4  Z_5   W_7 2023-07-05  0.333333     0.756930   
5233   ID_683  X_0  Y_2  Z_4   W_7 2023-11-12  1.000000     0.111138   

      Humidity  Wind_Speed  ...  C4_W_10  C4_W_11  C4_W_2  C4_W_3  C4_W_4  \
3919  0.946733    0.627192  ...    False    False   False    True   False   
3920  0.103760    0.888574  ...    False    False   False   False   False   
3921  0.342065    0.638082  ...     True    False   False   False   False   
3922  0.133882    0.565980  ...    False    False   False   False   False   
3923  0.835746    0.569353  ...    False    False   False   False   False   
...        ...         ...  ...      ...      ...     ...     ...     ...   
5229  0.735491    0.471118  ...    False    False   False   False   False   
5230  0.133882    0.565980  ...    False    False   False   False   False   
5231  0.902677    0.725309  ...    False    False   False   False   False   
5232  0.539271    0.684702  ...    False    False   False   False   False   
5233  0.164734    0.736248  ...    False    False   False   False   False   

      C4_W_5  C4_W_6  C4_W_7  C4_W_8  C4_W_9  
3919   False   False   False   False   False  
3920   False   False   False   False   False  
3921   False   False   False   False   False  
3922   False    True   False   False   False  
3923   False   False   False   False   False  
...      ...     ...     ...     ...     ...  
5229   False   False   False   False   False  
5230   False    True   False   False   False  
5231   False    True   False   False   False  
5232   False   False    True   False   False  
5233   False   False    True   False   False  

[1315 rows x 4279 columns]

In [419]:
from sklearn.model_selection import train_test_split
X_train, X_val, y_train, y_val = train_test_split(train, y, test_size=0.2, random_state=42, stratify=y)

In [420]:
X_train.shape
X_val.shape
y_train.shape
y_val.shape

(3135, 4279)

(784, 4279)

(3135,)

(784,)

In [ ]:
from sklearn.ensemble import RandomForest

# 다시 re

In [421]:
import pandas as pd

# 데이터 로드
train = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p2_train2.csv')
test = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p2_test2.csv')

# 타겟 분리
y = train['Y']
train.drop(columns=['Y'], inplace=True)

# ID 제거
train.drop(columns=['ID'], inplace=True)
test.drop(columns=['ID'], inplace=True)

# 날짜 처리 (파생 변수 생성)
for df in [train, test]:
    df['Date'] = pd.to_datetime(df['Date'])
    df['Year'] = df['Date'].dt.year
    df['Month'] = df['Date'].dt.month
    df['Day'] = df['Date'].dt.day
    df['DayofWeek'] = df['Date'].dt.dayofweek
    df.drop(columns=['Date'], inplace=True)

# 범주형 처리 (One-Hot Encoding)
categorical_cols = ['C1', 'C2', 'C3', 'C4', 'Weekday']
train = pd.get_dummies(train, columns=categorical_cols)
test = pd.get_dummies(test, columns=categorical_cols)

# (주의) test에만 있고 train에 없는 컬럼 또는 반대가 있을 수 있음 → align으로 컬럼 정렬
train, test = train.align(test, join='left', axis=1, fill_value=0)

# 이제 train, test는 모두 숫자형 컬럼만 포함된 상태


In [422]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.model_selection import train_test_split

# 학습용 분리
X_train, X_val, y_train, y_val = train_test_split(train, y, test_size=0.2, random_state=42)

# 모델 학습
model = RandomForestRegressor(random_state=42)
model.fit(X_train, y_train)

# 검증
pred = model.predict(X_val)
print("MAE:", mean_absolute_error(y_val, pred))


RandomForestRegressor(random_state=42)

MAE: 10.164777636594664


# 유형 3

url = https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p3_1.csv   
종속변수 : y  

In [2]:
import pandas as pd
df=pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p3_1.csv')

In [3]:
df.info()
df.isnull().sum()
df.nunique()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2500 entries, 0 to 2499
Data columns (total 16 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   X1      2500 non-null   float64
 1   X2      2500 non-null   float64
 2   X3      2500 non-null   float64
 3   X4      2500 non-null   float64
 4   X5      2500 non-null   float64
 5   X6      2500 non-null   float64
 6   X7      2500 non-null   float64
 7   X8      2500 non-null   float64
 8   X9      2500 non-null   float64
 9   X10     2500 non-null   float64
 10  X11     2500 non-null   float64
 11  X12     2500 non-null   float64
 12  X13     2500 non-null   float64
 13  X14     2500 non-null   float64
 14  X15     2500 non-null   float64
 15  Y       2500 non-null   int64  
dtypes: float64(15), int64(1)
memory usage: 312.6 KB


X1     0
X2     0
X3     0
X4     0
X5     0
X6     0
X7     0
X8     0
X9     0
X10    0
X11    0
X12    0
X13    0
X14    0
X15    0
Y      0
dtype: int64

X1     2500
X2     2500
X3     2500
X4     2500
X5     2500
X6     2500
X7     2500
X8     2500
X9     2500
X10    2500
X11    2500
X12    2500
X13    2500
X14    2500
X15    2500
Y         2
dtype: int64

In [15]:
df.head()

X1        X2        X3        X4        X5        X6        X7  \
0  0.506278 -0.231322  0.759771 -0.172927 -0.687524 -0.556815 -1.381369   
1  0.017220  1.530047  0.676838  1.533623 -0.516027  0.949166 -1.180669   
2  1.436089  0.038078 -0.120501  0.914955 -1.504149  0.443893 -0.696385   
3 -0.926337 -0.665242 -0.106853 -1.111627 -0.127867  0.746742 -0.345943   
4 -0.350953 -0.044709 -0.421777  1.359772 -1.619690 -0.645908 -1.717975   

         X8        X9       X10       X11       X12       X13       X14  \
0  1.633690 -0.869129  0.370614  0.615057 -0.416594 -1.175980  0.113494   
1  1.895021 -1.211768 -1.081109 -1.067039 -0.611829  0.846854 -0.812381   
2  0.245013  1.221166 -1.387287 -1.762048 -0.556466  0.130129  0.517210   
3  0.848144  0.374944 -0.951299  1.097359  0.754369 -0.221183 -0.430509   
4 -0.598477  1.026014 -0.134644 -0.988988  0.513048 -0.886113  0.247288   

        X15  Y  
0 -0.539230  1  
1  0.823546  0  
2  1.056730  0  
3  0.756875  1  
4 -0.016664  0

### 3-1) y컬럼을 종속변수로하는 로지스틱회귀 모형을 만들 떄 유의미하지 않은 변수의 개수는? % 상수항은 추가하여 모델링하라

In [4]:
y = df['Y']
train = df.drop('Y', axis=1)
y.head(2)
train.head(2)

0    1
1    0
Name: Y, dtype: int64

X1        X2        X3        X4        X5        X6        X7  \
0  0.506278 -0.231322  0.759771 -0.172927 -0.687524 -0.556815 -1.381369   
1  0.017220  1.530047  0.676838  1.533623 -0.516027  0.949166 -1.180669   

         X8        X9       X10       X11       X12       X13       X14  \
0  1.633690 -0.869129  0.370614  0.615057 -0.416594 -1.175980  0.113494   
1  1.895021 -1.211768 -1.081109 -1.067039 -0.611829  0.846854 -0.812381   

        X15  
0 -0.539230  
1  0.823546

In [7]:
# !pip install statsmodels

In [16]:
import statsmodels.api as sm

X = sm.add_constant(train)
model = sm.Logit(y, X)
result = model.fit()

print(result.summary())

Optimization terminated successfully.
         Current function value: 0.060586
         Iterations 10
                           Logit Regression Results                           
Dep. Variable:                      Y   No. Observations:                 2500
Model:                          Logit   Df Residuals:                     2484
Method:                           MLE   Df Model:                           15
Date:                Wed, 18 Jun 2025   Pseudo R-squ.:                  0.9126
Time:                        23:07:07   Log-Likelihood:                -151.47
converged:                       True   LL-Null:                       -1732.9
Covariance Type:            nonrobust   LLR p-value:                     0.000
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -2.1997      0.566     -3.883      0.000      -3.310      -1.089
X1             0.2990      0

In [30]:
# dir(result.summary())

In [33]:
dir(result)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_cache',
 '_data_attr',
 '_data_in_cache',
 '_get_endog_name',
 '_get_robustcov_results',
 '_get_wald_nonlinear',
 '_transform_predict_exog',
 '_use_t',
 'aic',
 'bic',
 'bse',
 'conf_int',
 'converged',
 'cov_kwds',
 'cov_params',
 'cov_type',
 'df_model',
 'df_resid',
 'f_test',
 'fittedvalues',
 'get_distribution',
 'get_influence',
 'get_margeff',
 'get_prediction',
 'im_ratio',
 'info_criteria',
 'initialize',
 'k_constant',
 'llf',
 'llnull',
 'llr',
 'llr_pvalue',
 'load',
 'method',
 'mle_retvals',
 'mle_settings',
 'model',
 'nobs',
 'normalized_cov_params',
 'params',
 'pred_table',
 'predict',

In [36]:
result.params
result.pvalues

const   -2.199743
X1       0.299010
X2       0.191273
X3       0.006576
X4       0.145479
X5       0.162473
X6       0.144967
X7      -1.254588
X8       0.772331
X9       0.149848
X10     -0.018957
X11      6.260117
X12     -0.082006
X13      0.054070
X14      0.073966
X15      0.243308
dtype: float64

const    1.031559e-04
X1       8.028705e-02
X2       2.944616e-01
X3       9.690616e-01
X4       3.911335e-01
X5       3.199229e-01
X6       3.958695e-01
X7       1.744923e-04
X8       7.251040e-07
X9       3.681983e-01
X10      9.148990e-01
X11      4.450390e-65
X12      6.205852e-01
X13      7.411058e-01
X14      6.567659e-01
X15      1.536596e-01
dtype: float64

In [41]:
# p-value 시리즈에서 0.05보다 큰 항목 수 세기 (상수항 제외)
p_values = result.pvalues.drop('const')
insignificant_count = (p_values > 0.05).sum()

print(f'유의하지 않은 변수 개수: {insignificant_count}')


유의하지 않은 변수 개수: 12


### 3-2) 유의한 변수들 만으로 다시 모델링을 해서 X를 포함하는 독립변수들의 회귀계수 합계를 소숫점 이하 3째 자리까지 구하여라 

In [66]:
import statsmodels.api as sm

# 상수항 추가
X = sm.add_constant(train)
model = sm.Logit(y, X)
result = model.fit()


Optimization terminated successfully.
         Current function value: 0.060586
         Iterations 10


In [67]:
df[p_values[p_values < 0.05].index]

X7        X8       X11
0    -1.381369  1.633690  0.615057
1    -1.180669  1.895021 -1.067039
2    -0.696385  0.245013 -1.762048
3    -0.345943  0.848144  1.097359
4    -1.717975 -0.598477 -0.988988
...        ...       ...       ...
2495 -1.614073  0.561380 -1.157305
2496 -0.586509  0.257907 -1.463763
2497 -1.066106  1.527682 -1.065597
2498 -0.853763  0.993235 -0.922527
2499 -0.995253  0.673438 -1.268256

[2500 rows x 3 columns]

In [69]:
# p-value < 0.05인 변수만 추출
significant_vars = result.pvalues[result.pvalues < 0.05].index

X_significant = X[significant_vars]
model2 = sm.Logit(y, X_significant)
result2 = model2.fit()
result2.summary()

Optimization terminated successfully.
         Current function value: 0.062467
         Iterations 10


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                      Y   No. Observations:                 2500
Model:                          Logit   Df Residuals:                     2496
Method:                           MLE   Df Model:                            3
Date:                Wed, 18 Jun 2025   Pseudo R-squ.:                  0.9099
Time:                        23:32:14   Log-Likelihood:                -156.17
converged:                       True   LL-Null:                       -1732.9
Covariance Type:            nonrobust   LLR p-value:                     0.000
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -2.0523      0.546     -3.756      0.000      -3.123      -0.981
X7            -1.1389      0.323     -3.525      0.000      -1.772      -0.506
X8             0.7529      0.149      5.041      0.000       0.460       1.046
X11            5.9960      0.326     18.374      0.000       5.356       6.636
==============================================================================
"""

In [70]:
coef = result2.params
coef

const   -2.052282
X7      -1.138880
X8       0.752897
X11      5.996007
dtype: float64

In [73]:
x_coef_sum = coef[coef.index.str.startswith('X')].sum()
round(x_coef_sum, 3)

5.61

In [74]:
# 계수 추출
coef = result2.params

# 'X'로 시작하는 변수만 필터링 후 합산
x_coef_sum = coef[coef.index.str.startswith('X')].sum()

# 소수점 이하 3째 자리까지 반올림
x_coef_sum_rounded = round(x_coef_sum, 3)

print(f"X로 시작하는 유의 변수들의 회귀계수 합계: {x_coef_sum_rounded}")


X로 시작하는 유의 변수들의 회귀계수 합계: 5.61


### 3-3) 3-2번의 모델 결과에 대해 X뒤의 숫자가 가장 낮은 컬럼이 5단위 늘어나면 오즈비 몇 배로 변화?

In [85]:
coef[coef.index.str.startswith('X')].index

Index(['X7', 'X8', 'X11'], dtype='object')

In [98]:
coef[coef.index.str.startswith('X')].index.sort_values

<bound method Index.sort_values of Index(['X7', 'X8', 'X11'], dtype='object')>

In [84]:
result2.pvalues

const    1.728738e-04
X7       4.232174e-04
X8       4.619437e-07
X11      2.131102e-75
dtype: float64

In [99]:
import re

# X로 시작하는 변수 추출
x_vars = coef[coef.index.str.startswith('X')].index

# 숫자 기준 정렬
sorted_vars = sorted(x_vars, key=lambda x: int(re.findall(r'\d+', x)[0]))

# 가장 숫자가 작은 변수
first_var = sorted_vars[0]
print(first_var)  # 예: 'X7'


X7


In [126]:
import numpy as np
coef
beta = coef['X7']
beta
odds = np.exp(beta * 5)
odds

const   -2.052282
X7      -1.138880
X8       0.752897
X11      5.996007
dtype: float64

-1.1388796733816182

0.0033647609218022748

In [122]:
coef
beta = coef[first_var]
beta
odds_ratio_change = np.exp(beta * 5)
odds_ratio_change
print(f"{first_var}이 5 증가하면 오즈비는 약 {odds_ratio_change:.3f}배 변화")


const   -2.052282
X7      -1.138880
X8       0.752897
X11      5.996007
dtype: float64

-1.1388796733816182

0.0033647609218022748

X7이 5 증가하면 오즈비는 약 0.003배 변화


데이터 설명 : 키,체중,뇌사이즈에 따른 지능(PIQ)   
데이터 출처 : online.stat.psu.edu  
데이터 url : https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p3_1.csv  

In [127]:
import pandas as pd 
df= pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p3_2.csv')
df.head()

PIQ   Brain  Height  Weight
0  124   81.69    64.5     118
1  150  103.84    73.3     143
2  128   96.54    68.8     172
3  134   95.15    65.0     147
4  110   92.88    69.0     146

In [128]:
df.info()
df.nunique()
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38 entries, 0 to 37
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   PIQ     38 non-null     int64  
 1   Brain   38 non-null     float64
 2   Height  38 non-null     float64
 3   Weight  38 non-null     int64  
dtypes: float64(2), int64(2)
memory usage: 1.3 KB


PIQ       21
Brain     38
Height    22
Weight    32
dtype: int64

PIQ       0
Brain     0
Height    0
Weight    0
dtype: int64

### 3-4) PIQ를 예측하는 다중선형회귀 모형을 만들고 Brain의 회귀 계수를 구하여라

In [129]:
import pandas as pd
import statsmodels.api as sm

# 데이터 불러오기
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/krdatacertificate/e8_p3_2.csv')

# 독립변수와 종속변수 분리
X = df[['Brain', 'Height', 'Weight']]
y = df['PIQ']

# 상수항 추가
X = sm.add_constant(X)

# 다중선형회귀 모형 적합
model = sm.OLS(y, X)
result = model.fit()

# Brain 변수의 회귀계수 출력
print("Brain 회귀계수:", result.params['Brain'])


Brain 회귀계수: 2.0603667978949485


### 3-5) 모델의 r2를 구하여라

### 3-6) 위에서 계산한 모델을 이용하여 뇌크기 90 키 70 몸무게 150일 때 지능을 예측하라